### PR TITLE
Update to 4.1.0

### PR DESCRIPTION
**Pull request by tamird**
@marc @bdarnell

### CODE REVIEW

File: import.sh [hunk 1]
```diff
@@ -4,8 +4,19 @@ set -eu
...
 rm -rf *.c internal/*
 curl -sL https://github.com/jemalloc/jemalloc/releases/download/4.0.4/jemalloc-4.0.4.tar.bz2 | tar jxf - -C internal --strip-components=1
-(cd internal && ./autogen.sh && ./configure --with-jemalloc-prefix="")
-patch -p1 < gitignore.patch
+
+# You need to manually run the following code.
+# on OSX:
+# mkdir -p darwin_includes/jemalloc/internal
+# (cd internal && ./autogen.sh && ./configure --enable-prof --with-jemalloc-prefix="")
+# cp internal/include/jemalloc/*.h darwin_includes/jemalloc
```
Comment (bdarnell) [inline import.sh]: Do you have to duplicate all the headers in the `$PLATFORM_includes` directory, or could you just put the generated ones there? All of the generated headers should have a corresponding `.in` file. (or maybe there's a `configure` argument to put the headers somewhere else, but if so it's not obvious)

**Comment by bdarnell** [comment]
LGTM

File: cgo_flags.go [hunk 2]
```diff
@@ -3,7 +3,7 @@
 package jemalloc
 
 // #cgo CFLAGS: -Iinternal/include
-// #cgo linux CFLAGS: -DJEMALLOC_PURGE_MADVISE_DONTNEED
-// #cgo darwin CFLAGS: -DJEMALLOC_OSATOMIC -DJEMALLOC_OSSPIN -DJEMALLOC_PURGE_MADVISE_FREE -DJEMALLOC_HAVE_ISSETUGID
-// #cgo darwin CFLAGS: -DJEMALLOC_ZONE -DJEMALLOC_ZONE_VERSION=8
+// #cgo darwin CFLAGS: -Idarwin_includes -Idarwin_includes/jemalloc/internal
+// #cgo linux CFLAGS: -Ilinux_includes -Ilinux_includes/jemalloc/internal
+// #cgo linux LDFLAGS: -lm
```
Comment (mberhault) [inline cgo_flags.go]: do you not need `-lm` on darwin? It was needed on linux after I enabled profiling and the `log()` calls in `prof.c` came into play

**Comment by mberhault** [comment]
LGTM

File: cgo_flags.go (hunk 2 above)
Comment (bdarnell) [inline cgo_flags.go]: On OSX, both `libc` and `libm` are provided by `libSystem` which is linked by default (http://stackoverflow.com/questions/33675638/gcc-link-the-math-library-by-default-in-c-on-mac-os-x)

But it's still a little more correct to pass `-lm` even on darwin so we should just remove the platform tag from this line.

**Comment by mberhault** [comment]
Seeing the errors on restrict still. It may be just my setup though:

```
../c-jemalloc/util.c:108:39: error: expected ‘;’, ‘,’ or ‘)’ before ‘nptr’
 malloc_strtoumax(const char *restrict nptr, char **restrict endptr, int base)
```

I think it's supposed to automatically `define` it if not supported, but we may be using different toolchains.

**Comment by tamird** [comment]
You're using an older compiler, see first commit.

File: import.sh (hunk 1 above)
Comment (tamird) [inline import.sh]: There are a number of generated files which do not have corresponding `.in` files - e.g. jemalloc.h.

I could use a whitelist, but it's more trouble than it's worth, I think.

I don't see a suitable flag to `configure`.

File: cgo_flags.go (hunk 2 above)
Comment (tamird) [inline cgo_flags.go]: Thanks for the explanation. Done.

File: import.sh (hunk 1 above)
Comment (tamird) [inline import.sh]: Ah, `git clean -n` to the rescue. Now copying the minimum number of headers. PTAL.

**Pull request merged**